xla/backends/profiler/gpu/cupti_tracer.h

```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+  std::vector<uint8_t> config_image_;
+  CUpti_PmSampling_Object* sampling_obj_ = nullptr;
+  std::vector<uint8_t> counter_data_image_;
+  std::unique_ptr< const std::vector<uint8_t> > p_counter_data_image_backup_;
+
+  // XLA interface to CUPTI
+  // Needed both to call PM sampling APIs and stringify CUPTI errors
+  CuptiInterface* cupti_interface_;
+
+  // Configuration calls
+  absl::Status GetChipName();
+  absl::Status DeviceSupported();
+  absl::Status CreateCounterAvailabilityImage();
+
+  // Requires counter availability image
+  absl::Status CreatProfilerHostObj();           
+
+  // Requires profiler host object
+  absl::Status CreateConfigImage();
+
+  // Requires config image
+  size_t NumPasses();
+
+  absl::Status InitializeProfilerAPIs();
+  absl::Status CreatePmSamplerObject();
+
+  // Requires pm sampler object
+  absl::Status CreateCounterDataImage();
+
+  // Clean up
+  absl::Status DestroyCounterAvailabilityImage();
+  absl::Status DestroyProfilerHostObj();
+  absl::Status DestroyConfigImage();
+  absl::Status DestroyPmSamplerObject();
+  absl::Status UnInitializeProfilerAPIs();
+  absl::Status DestroyCounterDataImage();
+
+  std::vector<char const*> default_metrics_{
+    "sm__cycles_active.sum",
+    "sm__inst_executed_pipe_fmalite.sum",
+    "pcie__read_bytes.sum",
+    "pcie__write_bytes.sum"
+  };
+
+  public:
+  // Device information
+  int device_id_;
+
+  // PM sampling public configuration
+  struct PmSamplingConfig* config_;
+
+  // Creates host and sampler objects, all images
+  absl::Status CreateConfig();
+
+  // Requires config image, pm sampler object
+  absl::Status SetConfig();
+
+  // Requires pm sampler object
+  absl::Status StartSampling();
+
+  // Requires pm sampler object
+  absl::Status StopSampling();
+
+  // Requires pm sampler object
+  absl::Status DisableSampling();
+
+  // Collect sampling data
+  // Requires pm sampler object, counter data image, fetches data from hw
+  // buffer into counter data image
+  absl::Status FillCounterDataImage(struct PmSamplingDecodeInfo& decodeInfo);
+
+  // Requires counter data image
+  absl::Status GetSampleCounts(struct PmSamplingDecodeInfo& decodeInfo);
+
+  // Requires host object, pm sampler object, counter data image, metric
+  // names, returns sample time, metric values
+  absl::Status GetSample(SamplerRange& sample, size_t index);
+
+  // Requires pm sampler object, counter data image, (re)initializes it
+  absl::Status InitializeCounterDataImage();
+
+  // Restores image from backup (faster than re-initializing)
+  absl::Status RestoreCounterDataImage();
+
+  // Constructor provides all configuration needed to set up sampling on a
+  // single device
+  PmSamplingDevice(int device_id, struct PmSamplingConfig* config);
+};
+
+// Container for PM sampling decode thread
+// Responsible for fetching PM sampling data from device and providing it to
+// handler or other external container
+class PmSamplingDecodeThread {
+  enum thd_state {
```

Comment:
Enumerators  should be named like constants: https://google.github.io/styleguide/cppguide.html#Enumerator_Names